auth/credentials.py

```diff
@@ -83,10 +120,100 @@ def validate_credentials(auth_username, auth_password_or_token):
         logger.debug("Found credentials header for robot %s", auth_username)
         try:
             robot = model.user.verify_robot(auth_username, auth_password_or_token)
+
             logger.debug("Successfully validated credentials for robot %s", auth_username)
+
             return ValidateResult(AuthKind.credentials, robot=robot), CredentialKind.robot
-        except model.InvalidRobotException as ire:
+        except model.DeactivatedRobotOwnerException as dre:
+            robot_owner, robot_name = parse_robot_username(auth_username)
+
+            try:
+                performer = model.user.lookup_robot(auth_username)
+            except User.DoesNotExist:
+                performer = None
+
+            logger.debug(
+                "Tried to use the robot %s for a disabled user: %s", (robot_name, robot_owner)
+            )
+
+            if app.config.get("ACTION_LOG_AUDIT_LOGIN_FAILURES"):
+                log_action(
+                    "login_failure",
+                    robot_owner,
+                    {
+                        "type": "v2auth",
+                        "kind": "robot",
+                        "robot": auth_username,
+                        "username": robot_owner,
+                        "useragent": request.user_agent.string,
+                        "message": str(dre),
+                    },
+                    performer=performer,
+                )
+
+            return (
+                ValidateResult(AuthKind.credentials, error_message=str(dre)),
+                CredentialKind.robot,
+            )
+        except model.InvalidRobotCredentialException as ire:
             logger.warning("Failed to validate credentials for robot %s: %s", auth_username, ire)
+
+            robot_owner, _ = parse_robot_username(auth_username)
+
+            try:
+                performer = model.user.lookup_robot(auth_username)
```

Comment:
Is this adding a DB call on failed logins that's not behind a feature flag? Same for the other DB calls in the other exception cases.

Reply:
it looks like they should be in the `if app.config.get("ACTION_LOG_AUDIT_LOGIN_FAILURES"):` block